### PR TITLE
[WIP] Estimator comparison

### PR DESCRIPTION
**Pull request by maxentile**
Machinery for comparing the near-equilibrium estimator vs nested monte carlo estimators of an exact expression for the KL divergence. Since there is monte carlo in both an inner loop and outer loop, the second estimator will in general be biased unless the number of inner loop samples goes to infinity. However, the extent of this bias may be small in most cases. Implementing John's offline suggestions, this supports adaptive allocation of computational effort to meet a target uncertainty threshold on either loop independently, which should hopefully achieve a good trade-off between expense and bias.

Also contains some of the fixes / improvements that were sitting around on other branches or locally (addressing box vector bug, storing samples in .h5 instead of .npy, addressing unit bug).

Before review, I still want to:
- [x] Add my LSF scripts
- [ ] Add the plotting / post-processing code from my notebooks
- [x] Improve documentation
- [x] Refactor the repeated structure in the adaptive loops

### CODE REVIEW

File: benchmark/evaluation/compare_near_eq_and_exact.py [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+
+    return {
+        "xv": (x, v),
+        "Ws": Ws,
+        "estimate": estimate_from_work_samples(Ws),
+        "W_shad_forward": W_shad_forward
+    }
+
+
+def noneq_sim_factory(testsystem_name, scheme, dt, collision_rate):
```
Comment (jchodera) [inline benchmark/evaluation/compare_near_eq_and_exact.py]: In cases where some arguments are bearing units and some are unitless (why would you do that???) it's CRITICAL to document your parameters and check that the user hasn't mis-specified any that should or should not bear units. It's important to remember that OpenMM doesn't check for unit compatibility, so it's easy to make mistakes like passing `dt = 1  * unit.femtosecond` and then this becomes `1 * unit.femtosecond**2` and would be stripped of units to be much smaller than you intended.

File: benchmark/evaluation/compare_near_eq_and_exact.py [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+
+def process_outer_samples(outer_samples):
+    """Compute near-equilibrium and "exact" estimates of D_KL from work samples"""
+
+    # estimate D_KL as a sample average over x ~ rho of log(rho(x) / pi(x))
+    new_estimate = np.mean([s["estimate"] for s in outer_samples], 0)
+
+    # near-equilibrium estimate
+    W_F_hat = np.mean([s["W_shad_forward"] for s in outer_samples])
+    W_R_hat = np.mean([s["Ws"][0] for s in outer_samples])  # picking the first W_R sample associated with each W_F
```
Comment (jchodera) [inline benchmark/evaluation/compare_near_eq_and_exact.py]: It's not obvious to me that this is the work accumulated over the second half of the trajectory (after velocity randomization). How can I figure out what this is doing?

**Comment by maxentile** [comment]
Thanks for these notes, John -- I'll take a stab at addressing these today

File: benchmark/evaluation/compare_near_eq_and_exact.py [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+marginals = ["configuration", "full"]
+dt_range = np.array([0.1] + list(np.arange(0.5, 4.001, 0.5)))
+
+# constant parameters
+collision_rate = 1.0 / unit.picoseconds
+temperature = simulation_parameters['temperature']
+n_steps = 1000  # number of steps until system is judged to have reached "steady-state"
+
+# naive inner-loop params
+n_inner_samples = 100
```
[the remark this answers is not in the document]
Comment (maxentile) [reply benchmark/evaluation/compare_near_eq_and_exact.py]: Addressed in https://github.com/choderalab/integrator-benchmark/pull/45/commits/71302bed96994a28a2e3c1570ff076e82c6dd16c

File: benchmark/evaluation/compare_near_eq_and_exact.py [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+
+# adaptive inner-loop params
+inner_loop_initial_size = 50
+inner_loop_batch_size = 1
+inner_loop_stdev_threshold = 0.1
+inner_loop_max_samples = 1000
+
+# naive outer-loop params
+n_outer_samples = 100
```
[the remark this answers is not in the document]
Comment (maxentile) [reply benchmark/evaluation/compare_near_eq_and_exact.py]: Addressed in https://github.com/choderalab/integrator-benchmark/pull/45/commits/a69d50c4f3bd67bc1498a9e5b87f0374ae42c89b

File: benchmark/evaluation/compare_near_eq_and_exact.py [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+
+# adaptive outer-loop params
+outer_loop_initial_size = 50
+outer_loop_batch_size = 1
+outer_loop_stdev_threshold = 0.1
+outer_loop_max_samples = 1000
+
+
+def stdev_log_rho_pi(w):
+    """Approximate the standard deviation of the estimate of log < e^{-w} >_{x; \Lambda}
```
[the remark this answers is not in the document]
Comment (maxentile) [reply benchmark/evaluation/compare_near_eq_and_exact.py]: Mostly addressed in https://github.com/choderalab/integrator-benchmark/pull/45/commits/39e07ff4e552aa9fdd6a679ff905f1aca8e05276 and https://github.com/choderalab/integrator-benchmark/pull/45/commits/1df681666c5aa41f6fc9e00435a6d817e61d501d . Will consolidate these functions in next PR

File: benchmark/evaluation/compare_near_eq_and_exact.py [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+def stdev_kl_div(outer_samples):
+    """Approximate the stdev of the estimate of E_rho log_rho_pi"""
+
+    # TODO: Propagate uncertainty from the estimates of log_rho_pi
+    # currently, just use standard error of mean of log_rho_pi_s
+    log_rho_pi_s = np.array([np.log(np.mean(np.exp(-sample['Ws']))) for sample in outer_samples])
+    return np.std(log_rho_pi_s) / np.sqrt(len(log_rho_pi_s))
+
+
+def estimate_from_work_samples(work_samples):
```
[the remark this answers is not in the document]
Comment (maxentile) [reply benchmark/evaluation/compare_near_eq_and_exact.py]: Will consolidate in next PR

File: benchmark/evaluation/compare_near_eq_and_exact.py (hunk 1 above)
Comment (maxentile) [reply benchmark/evaluation/compare_near_eq_and_exact.py]: Addressed in https://github.com/choderalab/integrator-benchmark/pull/45/commits/3145e26ff0803b4fd3632866d0d8c147c3e41c34 (The unitless dt issue was a hack around a bug i had pickling / unpickling the "experiment" tuple, but it was unnecessary.)

File: benchmark/evaluation/compare_near_eq_and_exact.py (hunk 2 above)
Comment (maxentile) [reply benchmark/evaluation/compare_near_eq_and_exact.py]: `outer_samples` is a list of dicts, and each element's `Ws` item is a list of inner-loop work samples from the second half of the trajectory, and I pick an arbitrary sample from that. I could have replaced `[0]` in L183 with `[np.random.randint(len(s['Ws']))]`, for example

**Comment by maxentile** [comment]
Merging now, and will add the analysis and plotting notebooks in a future PR when all the simulation results are available.

**Pull request merged**